### PR TITLE
refactor: Updated `@google/genai` instrumentation to subscribe to events emitted

### LINKED ISSUE #3441: Update google genai instrumentation to rely on tracing channel instrumentation
### Description
Update instrumentation to create a subscriber for google genai and a instrumentator for orchestrion.  

### Acceptance Criteria
google genai instrumentation still functions as expected after traditional monkey patching and shim helpers are used.

### PR DESCRIPTION
**Pull request by amychisholm03**
This PR resolves #3441

### CODE REVIEW

File: lib/subscribers/google-genai/generate-content-stream.js [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+
+  /**
+   * Instruments the streaming response by wrapping the next function.
+   * @param {object} params function params
+   * @param {object} params.request the original request object
+   * @param {object} params.response the original response object
+   * @param {TraceSegment} params.segment the active trace segment
+   * @param {Transaction} params.transaction the active transaction
+   */
+  async instrumentStream({ request, response, segment, transaction }) {
```
Comment (bizob2828) [inline lib/subscribers/google-genai/generate-content-stream.js]: this isn't an async method.

File: lib/subscribers/google-genai/generate-content-stream.js [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+    let content
+    let modelVersion
+    let finishReason
+    let entireMessage = ''
+    response.next = async function wrappedNext(...nextArgs) {
+      let result = {}
+      try {
+        result = await originalNext.apply(response, nextArgs)
+        if (result?.value?.text) {
+          modelVersion = result.value.modelVersion
```
Comment (bizob2828) [inline lib/subscribers/google-genai/generate-content-stream.js]: you're using optional chaining right below but not here, any reason?

File: lib/subscribers/google-genai/generate-content-stream.js [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+            segment,
+            transaction,
+            request,
+            response: result?.value,
+            err
+          })
+        }
+      }
+      return result
+    }.bind(this)
```
Comment (bizob2828) [inline lib/subscribers/google-genai/generate-content-stream.js]: you _could_ just save off `this` references as `const self = this` and use `self.<fn>` within the wrapped stream

File: lib/subscribers/google-genai/generate-content-stream.js [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+    if (!agent.config.ai_monitoring.streaming.enabled) {
+      logger.warn(
+        '`ai_monitoring.streaming.enabled` is set to `false`, stream will not be instrumented.'
+      )
+      agent.metrics.getOrCreateMetric(AI.STREAMING_DISABLED).incrementCallCount()
+      return
+    }
+
+    const originalNext = response.next
+    let err
```
Comment (bizob2828) [inline lib/subscribers/google-genai/generate-content-stream.js:44]: any reason why these variables are defined outside the scope of the wrapped function? They should be specific to every invocation. it's fine because this is wrapped on every call but just a little peculiar

File: lib/subscribers/google-genai/generate-content-stream.js [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+  asyncEnd(data) {
+    const ctx = this.agent.tracer.getContext()
+    if (!ctx?.segment || !ctx?.transaction) {
+      return
+    }
+    const { result: response, arguments: args, error: err } = data
+    const [request] = args
+    this.instrumentStream({
+      request,
+      headers: ctx.extras?.headers,
```
Comment (bizob2828) [inline lib/subscribers/google-genai/generate-content-stream.js]: this isn't even getting set nor used. I assume this is copy pasta from openai

File: lib/subscribers/google-genai/generate-content.js [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+    }
+    const { result: response, arguments: args, error: err } = data
+    const [request] = args
+
+    this.recordChatCompletionMessages({
+      segment: ctx.segment,
+      transaction: ctx.transaction,
+      request,
+      response,
+      headers: ctx.extras?.headers,
```
Comment (bizob2828) [inline lib/subscribers/google-genai/generate-content.js]: this isn't even getting set nor used. I assume this is copy pasta from openai

File: test/unit/instrumentation/@google/genai.test.js [hunk 7]
```diff
@@ -1,96 +0,0 @@
...
-        assert.equal(
-          shim.logger.warn.args[0][0],
-          '`ai_monitoring.streaming.enabled` is set to `false`, stream will not be instrumented.'
-        )
-        end()
-      })
-    }
-  )
-
-  await t.test(
```
Comment (bizob2828) [inline test/unit/instrumentation/@google/genai.test.js:77]: we don't have a test for this now anywhere

File: lib/subscribers/google-genai/generate-content-stream.js [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+      response,
+      segment: ctx.segment,
+      transaction: ctx.transaction,
+      err
```
Comment (bizob2828) [inline lib/subscribers/google-genai/generate-content-stream.js]: I don't see this being used

File: lib/subscribers/google-genai/generate-content-stream.js [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+   * @param {object} params function params
+   * @param {object} params.request the original request object
+   * @param {object} params.response the original response object
+   * @param {TraceSegment} params.segment the active trace segment
+   * @param {Transaction} params.transaction the active transaction
+   */
+  async instrumentStream({ request, response, segment, transaction }) {
+    const { agent, logger } = this
+
+    if (!agent.config.ai_monitoring.streaming.enabled) {
```
Comment (bizob2828) [inline lib/subscribers/google-genai/generate-content-stream.js]: this check is too late. it has already created the segment because you're inheriting from GoogleGenAISubscriber. In the previous instrumentation it was wrapping the stream method but not creating the segment because a recorder spec wasn't getting returned. You need to override the handler method and check that there, otherwise return the current ctx

File: lib/subscribers/google-genai/embed-content.js [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+  }
+
+  asyncEnd(data) {
+    const agent = this.agent
+    const ctx = agent.tracer.getContext()
+    if (!ctx?.segment || !ctx?.transaction) {
+      return
+    }
+    this.addLlmMeta({ transaction: ctx.transaction, version: data.moduleVersion })
+    let { result: response, arguments: args, error: err } = data
```
Comment (bizob2828) [inline lib/subscribers/google-genai/embed-content.js]:
```suggestion
     // If we get an error, it is possible that `response = null`.
     // In that case, we define it to be an empty object.
    const { result: response = {}, arguments: args, error: err } = data
```

File: lib/subscribers/google-genai/embed-content.js [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+    const [request] = args
+
+    if (!response) {
+      // If we get an error, it is possible that `response = null`.
+      // In that case, we define it to be an empty object.
+      response = {}
+    }
```
Comment (bizob2828) [inline lib/subscribers/google-genai/embed-content.js]:
```suggestion
```

File: lib/subscribers/google-genai/generate-content-stream.js [hunk 12]
```diff
@@ -0,0 +1,113 @@
...
+      return result
+    }.bind(this)
+  }
+
+  asyncEnd(data) {
+    const ctx = this.agent.tracer.getContext()
+    if (!ctx?.segment || !ctx?.transaction) {
+      return
+    }
+    const { result: response, arguments: args, error: err } = data
```
Comment (bizob2828) [inline lib/subscribers/google-genai/generate-content-stream.js]: should this also set `response` to `response = {}`? Are we missing a test like embedding where if it errors response it null?

File: lib/subscribers/google-genai/base.js [hunk 13]
```diff
@@ -0,0 +1,127 @@
...
+      withError: err != null
+    })
+
+    // Only take the first response message and append to input messages
+    // request.contents can be a string or an array of strings
+    // response.candidates is an array of candidates (choices); we only take the first one
+    const inputMessages = Array.isArray(request.contents) ? request.contents : [request.contents]
+    const responseMessage = response?.candidates?.[0]?.content
+    const messages = responseMessage !== undefined ? [...inputMessages, responseMessage] : inputMessages
+    messages.forEach((message, index) => {
```
Comment (bizob2828) [inline lib/subscribers/google-genai/base.js]: This was refactored to remove .`forEach`, [see](https://github.com/newrelic/node-newrelic/blob/f281f19fbc40f91ae80957ef8c6b98f1df23f606/lib/instrumentation/%40google/genai.js#L98)

File: lib/subscribers/google-genai/base.js [hunk 14]
```diff
@@ -0,0 +1,127 @@
...
+   * @param {TraceSegment} params.segment active segment from chat completion
+   * @param {object} params.request chat completion params
+   * @param {object} [params.response] chat completion response
+   * @param {boolean} [params.err] err if it exists
+   * @param {Transaction} params.transaction active transaction
+   */
+  recordChatCompletionMessages({
+    segment,
+    request,
+    response,
```
Comment (bizob2828) [inline lib/subscribers/google-genai/base.js]:
```suggestion
    response = {},
```

File: lib/subscribers/google-genai/base.js [hunk 15]
```diff
@@ -0,0 +1,127 @@
...
+    err,
+    transaction
+  }) {
+    const agent = this.agent
+    if (!response) {
+      // If we get an error, it is possible that `response = null`.
+      // In that case, we define it to be an empty object.
+      response = {}
+    }
```
Comment (bizob2828) [inline lib/subscribers/google-genai/base.js]:
```suggestion
```

File: lib/subscribers/google-genai/config.js [hunk 16]
```diff
@@ -0,0 +1,84 @@
...
+ * Copyright 2025 New Relic Corporation. All rights reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+const generateContentInternal = {
+  path: './google-genai/generate-content.js',
+  instrumentations: [
+    {
+      channelName: 'nr_generateContentInternal',
+      module: { name: '@google/genai', versionRange: '>=1.1.0 <1.5.0 || >=1.5.1', filePath: 'dist/node/index.cjs' },
```
Comment (bizob2828) [inline lib/subscribers/google-genai/config.js]: shouldn't this just be `>=1.1.0`? I get 1.5.0 broke our tests but the structures of objects should be the same, right?

File: lib/subscribers/google-genai/embed-content.js [hunk 17]
```diff
@@ -0,0 +1,66 @@
...
+const GoogleGenAISubscriber = require('./base')
+const { AI } = require('../../../lib/metrics/names')
+const { GEMINI } = AI
+const {
+  LlmErrorMessage,
+  LlmEmbedding
+} = require('../../../lib/llm-events/google-genai')
+
+class GoogleGenAIEmbedContentSubscriber extends GoogleGenAISubscriber {
+  constructor ({ agent, logger, channelName = 'nr_embedContent' }) {
```
Comment (bizob2828) [inline lib/subscribers/google-genai/embed-content.js]: I'd just assign channelName when you call super

File: lib/subscribers/google-genai/generate-content-stream.js [hunk 18]
```diff
@@ -0,0 +1,113 @@
+/*
+ * Copyright 2025 New Relic Corporation. All rights reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+const GoogleGenAIGenerateContentSubscriber = require('./generate-content')
+const { AI } = require('../../../lib/metrics/names')
+
+class GoogleGenAIGenerateContentStreamSubscriber extends GoogleGenAIGenerateContentSubscriber {
+  constructor({ agent, logger, channelName = 'nr_generateContentStreamInternal' }) {
```
Comment (bizob2828) [inline lib/subscribers/google-genai/generate-content-stream.js]: I'd just assign channelName when you call super

**Review by bizob2828 — CHANGES_REQUESTED** [review]
Nice work, I had a few comments around some things. I think we can also file a ticket to abstract LLM based instrumentation to have a base LLM class to remove this copy pasta around recording messages, errors, etc but for a different story.

File: lib/subscribers/google-genai/generate-content-stream.js (hunk 4 above)
Comment (amychisholm03) [reply lib/subscribers/google-genai/generate-content-stream.js:44]: It will fail if these variables are declared within `response.next` because not all chunks will have model version or stop reason, so we need to store them so they don't get overridden. `entireMessage` concats the entire response as text and will not work if it is being reset to `''` every time a new chunk comes in.

File: lib/subscribers/google-genai/generate-content-stream.js (hunk 2 above)
Comment (amychisholm03) [reply lib/subscribers/google-genai/generate-content-stream.js]: No reason, just copy and pasted from old code.

File: lib/subscribers/google-genai/generate-content-stream.js (hunk 5 above)
Comment (amychisholm03) [reply lib/subscribers/google-genai/generate-content-stream.js]: Yup, I'll remove it

File: lib/subscribers/google-genai/generate-content-stream.js (hunk 9 above)
Comment (bizob2828) [reply lib/subscribers/google-genai/generate-content-stream.js]: this comment wasn't addressed

Comment (bizob2828) [reply lib/subscribers/google-genai/generate-content-stream.js]: the test where it sets streaming disable should assert there isn't a segment getting created for `Llm/completion/Gemini/generate_content`, if that assertion was there, it would now be failing

Comment (amychisholm03) [reply lib/subscribers/google-genai/generate-content-stream.js]: Thanks, looking into it now; a lot of these comments I accidentally hit hide/resolve

File: lib/subscribers/google-genai/generate-content-stream.js [hunk 19]
```diff
@@ -0,0 +1,118 @@
...
+            err
+          })
+        }
+      }
+      return result
+    }.bind(this)
+  }
+
+  asyncEnd(data) {
+    if (!this.agent?.config?.ai_monitoring?.streaming?.enabled) {
```
Comment (bizob2828) [inline lib/subscribers/google-genai/generate-content-stream.js]: don't you want this to also go up in the handler? I know you'd still have to check if it was disabled to not wrap the stream but incrementing the metric may make more sense above

File: lib/subscribers/google-genai/base.js [hunk 20]
```diff
@@ -0,0 +1,123 @@
...
+const Subscriber = require('../base')
+
+class GoogleGenAISubscriber extends Subscriber {
+  constructor({ agent, logger, channelName, packageName = '@google/genai' }) {
+    super({ agent, logger, channelName, packageName })
+    this.events = ['asyncEnd']
+  }
+
+  get enabled() {
+    return super.enabled && this.agent?.config?.ai_monitoring?.enabled
```
Comment (bizob2828) [inline lib/subscribers/google-genai/base.js]:
```suggestion
    return super.enabled && this.?.config?.ai_monitoring?.enabled
```

File: lib/subscribers/google-genai/generate-content-stream.js [hunk 21]
```diff
@@ -0,0 +1,118 @@
...
+const GoogleGenAIGenerateContentSubscriber = require('./generate-content')
+const { AI } = require('../../../lib/metrics/names')
+
+class GoogleGenAIGenerateContentStreamSubscriber extends GoogleGenAIGenerateContentSubscriber {
+  constructor({ agent, logger }) {
+    super({ agent, logger, channelName: 'nr_generateContentStreamInternal' })
+  }
+
+  handler(data, ctx) {
+    if (!this.agent?.config?.ai_monitoring?.streaming?.enabled) {
```
Comment (bizob2828) [inline lib/subscribers/google-genai/generate-content-stream.js]:
```suggestion
    if (!this?.config?.ai_monitoring?.streaming?.enabled) {
```

File: lib/subscribers/google-genai/generate-content.js [hunk 22]
```diff
@@ -0,0 +1,54 @@
...
+const { AI } = require('../../metrics/names')
+const { GEMINI } = AI
+
+class GoogleGenAIGenerateContentSubscriber extends GoogleGenAISubscriber {
+  constructor({ agent, logger, channelName = 'nr_generateContentInternal' }) {
+    super({ agent, logger, channelName })
+  }
+
+  handler(data, ctx) {
+    if (!this.agent?.config?.ai_monitoring?.enabled) {
```
Comment (bizob2828) [inline lib/subscribers/google-genai/generate-content.js]:
```suggestion
    if (!this?.config?.ai_monitoring?.enabled) {
```

Comment (bizob2828) [reply lib/subscribers/google-genai/generate-content.js]: I'm confused why is this being checked? It's build into the google gen ai base handler?

Comment (bizob2828) [reply lib/subscribers/google-genai/generate-content.js]: OIC, this is because you can dynamically set this config after start up.  I missed that when porting over openai. We should clean this up to reduce redundancy though but we can handle that in the AIM subscriber abstraction. Can you make sure you file that ticket?

File: lib/subscribers/google-genai/generate-content.js [hunk 23]
```diff
@@ -0,0 +1,54 @@
...
+    const segment = this.agent.tracer.createSegment({
+      name: GEMINI.COMPLETION,
+      parent: ctx.segment,
+      transaction: ctx.transaction
+    })
+    return ctx.enterSegment({ segment })
+  }
+
+  asyncEnd(data) {
+    if (!this.agent?.config?.ai_monitoring?.enabled) {
```
Comment (bizob2828) [inline lib/subscribers/google-genai/generate-content.js]:
```suggestion
    if (!this?.config?.ai_monitoring?.enabled) {
```

File: lib/subscribers/google-genai/generate-content.js (hunk 22 above)
Comment (amychisholm03) [reply lib/subscribers/google-genai/generate-content.js]: Yes I was planning on making the ticket shortly

File: lib/subscribers/google-genai/generate-content-stream.js (hunk 19 above)
Comment (amychisholm03) [reply lib/subscribers/google-genai/generate-content-stream.js]: I think it should be in one place or the other. Putting it in both might increment the metric twice.

**Comment by amychisholm03** [comment]
`agent.config` -> `config` will be addressed in #3487

File: lib/subscribers/google-genai/generate-content-stream.js (hunk 21 above)
Comment (bizob2828) [reply lib/subscribers/google-genai/generate-content-stream.js]: why was this resolved without being applied? Yes `this.agent.config` works but we're trying to standardize using `this.config`

File: lib/subscribers/google-genai/generate-content-stream.js (hunk 19 above)
Comment (bizob2828) [reply lib/subscribers/google-genai/generate-content-stream.js]: agreed but it's still in 2 places. you should keep it only in the handler

File: lib/subscribers/google-genai/base.js (hunk 20 above)
Comment (bizob2828) [inline lib/subscribers/google-genai/base.js]: remove optional chaining this value will always be present

Comment (amychisholm03) [reply lib/subscribers/google-genai/base.js]: I don't know if `this.enabled` is the best way to check if ai monitoring is enabled because the log message that gets written specifically mentions that config value. It may be hard to debug. According to the [spec](https://source.datanerd.us/agents/agent-specs/blob/main/artificial-intelligence/LLMs.md#configuration), segments should not be created if ai_monitoring is disabled: "`ai_monitoring.enabled` indicates whether LLM instrumentation will be registered. If this is set to False, no metrics, events, or spans are to be sent."

File: lib/subscribers/google-genai/generate-content-stream.js [hunk 24]
```diff
@@ -0,0 +1,124 @@
...
+        // time it took to handle the stream
+        segment.touch()
+
+        // also need to enter this block if there was an
+        // error, so we can record it
+        if (isDone || err) {
+          if (cachedResult?.candidates?.[0]?.content?.parts) {
+            cachedResult.candidates[0].content.parts[0].text = entireMessage
+          }
+          this.recordChatCompletionMessages({
```
Comment (bizob2828) [inline lib/subscribers/google-genai/generate-content-stream.js]: this is confusing that `this` in here is the subscriber. I would remove the `.bind` and just save off this as `self`. I commented on this before but didn't see it and this threw me off again
